cabal: Add a flag to enable/disable scrypt

rvl opened the pull request:
### Issue Number

ADP-842
Fixes #2578.

### Overview

- Adds cabal flag so that cardano-wallet can be compiled without the `scrypt` package.
- Replaces `scrypt` with `cryptonite` for the integration tests only

If `scrypt` does not compile for your target system, then disable it with:

```
cabal configure --disable-tests --enable-benchmarks -f-scrypt
```

### Comments

I'm not too keen on changing the scrypt package in the main code, for the time being. But we can replace this package in the integration tests, without too much risk.

Need to use conditional modules instead of CPP because that breaks stylish-haskell.

rvl committed:
cabal: Add a flag to enable/disable scrypt

HeinrichApfelmus commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -1248,17 +1250,19 @@ emptyRandomWalletWithPasswd ctx rawPwd = do
         . CBOR.encodeBytes
         . BA.convert
 
+encryptPasswordWithScrypt :: (BA.ByteArrayAccess a, BA.ByteArray b) => a -> IO b
```
For posterity: The documentation of the `scrypt` package notes that the ByteString returned by `Scrypt.getEncryptedPass` contains not just the encrypted password, but also the parameters and salt used to encrypt it. In contrast, the `generate` function from the `cryptonite` package only seems to return that encrypted password.

The difference does not matter for the integration test, but it would matter for legacy wallets (hence the choice to keep the `scrypt` dependency in the core library.)

rvl replied:
Yes good point - thanks - this will need fixing.

HeinrichApfelmus reviewed (approved):
Looks good to me!

rvl committed:
Replace scrypt with cryptonite for tests only

I'm not too keen on changing the scrypt package in the main code, for
the time being. But we can replace this package in the integration
tests, without too much risk.

rvl committed:
Fixes

hamishmack committed:
Fix conversion error

hamishmack committed:
Turn off scrypt for aarch64

hamishmack committed:
Fix missing install_name_tool and work around code signing issue

hamishmack committed:
Fixes for cardano-wallet-core-integration

Commit:
Regenerate nix

Anviking committed:
Rebase and warning fixups

Anviking committed:
Adapt passphrase tests to work without scrypt

I simply moved them to LegacySpec, and ensured LegacySpec only ran when
scrypt was set.

However, I'm seeing the following failures with +scrypt set:

  test/unit/Cardano/Wallet/Primitive/Passphrase/LegacySpec.hs:53:9:
  1) Cardano.Wallet.Primitive.Passphrase.Legacy, Scrypt tests-only cryptonite version, Verify passphrase
       expected: True
        but got: False

  To rerun use: --match "/Cardano.Wallet.Primitive.Passphrase.Legacy/Scrypt tests-only cryptonite version/Verify passphrase/"

  test/unit/Cardano/Wallet/Primitive/Passphrase/LegacySpec.hs:68:9:
  2) Cardano.Wallet.Primitive.Passphrase.Legacy, Scrypt tests-only cryptonite version, getSalt
       expected: Just "abc"
        but got: Just "+jottFIZjVmXoC7LlTBZWlYzBX4QQwfduFuxCkMHUis="

  To rerun use: --match "/Cardano.Wallet.Primitive.Passphrase.Legacy/Scrypt tests-only cryptonite version/getSalt/"

  test/unit/Cardano/Wallet/Primitive/Passphrase/LegacySpec.hs:84:13:
  3) Cardano.Wallet.Primitive.Passphrase.Legacy, Scrypt tests-only cryptonite version, golden test legacy passphrase encryption, compare new implementation with cardano-sl - short password
       expected: Right ()
        but got: Left ErrWrongPassphrase

  To rerun use: --match "/Cardano.Wallet.Primitive.Passphrase.Legacy/Scrypt tests-only cryptonite version/golden test legacy passphrase encryption/compare new implementation with cardano-sl - short password/"

  test/unit/Cardano/Wallet/Primitive/Passphrase/LegacySpec.hs:94:13:
  4) Cardano.Wallet.Primitive.Passphrase.Legacy, Scrypt tests-only cryptonite version, golden test legacy passphrase encryption, compare new implementation with cardano-sl - normal password
       expected: Right ()
        but got: Left ErrWrongPassphrase

  To rerun use: --match "/Cardano.Wallet.Primitive.Passphrase.Legacy/Scrypt tests-only cryptonite version/golden test legacy passphrase encryption/compare new implementation with cardano-sl - normal password/"

  test/unit/Cardano/Wallet/Primitive/Passphrase/LegacySpec.hs:103:13:
  5) Cardano.Wallet.Primitive.Passphrase.Legacy, Scrypt tests-only cryptonite version, golden test legacy passphrase encryption, compare new implementation with cardano-sl - empty password
       expected: Right ()
        but got: Left ErrWrongPassphrase

  To rerun use: --match "/Cardano.Wallet.Primitive.Passphrase.Legacy/Scrypt tests-only cryptonite version/golden test legacy passphrase encryption/compare new implementation with cardano-sl - empty password/"

  test/unit/Cardano/Wallet/Primitive/Passphrase/LegacySpec.hs:113:13:
  6) Cardano.Wallet.Primitive.Passphrase.Legacy, Scrypt tests-only cryptonite version, golden test legacy passphrase encryption, compare new implementation with cardano-sl - cardano-wallet password
       expected: Right ()
        but got: Left ErrWrongPassphrase

  To rerun use: --match "/Cardano.Wallet.Primitive.Passphrase.Legacy/Scrypt tests-only cryptonite version/golden test legacy passphrase encryption/compare new implementation with cardano-sl - cardano-wallet password/"

Randomized with seed 866879004

Finished in 28.5483 seconds, used 40.2343 seconds of CPU time
85 examples, 6 failures

Anviking committed:
Correct Passphrase.Legacy implementation to make tests pass

1. Swap order of cborify and hashMaybe

This makes the legacy golden tests pass, and matches the old
implementation.

2. Fix getSalt implementation

The salt should be converted from base64. This fixes the test
"Verify passphrase" calling `checkPassphraseTestingOnly`.

The getSalt golden value was updated. With the "Verify passphrase" test
now succeeding, it seems like the previous expected salt value — "abc" —
was just a dummy value, and not the right expectation.

Anviking committed:
Tweaks to PassphraseSpec

Anviking commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -275,6 +275,8 @@ import qualified Cardano.Wallet.DB.MVar as MVar
 import qualified Cardano.Wallet.DB.Sqlite.AddressBook as Sqlite
 import qualified Cardano.Wallet.Primitive.Migration as Migration
+import Cardano.Wallet.Primitive.Passphrase.Current
+    ( preparePassphrase )
```
👀

Anviking committed:
Move cborify outside preparePassphrase

This way it matches the implemetnation on `master`.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4048,6 +4061,12 @@ instance IsServerError ErrWithRootKey where
+                [ "This build is not compiled with support for the "
+                , toText s <> " scheme used by the given wallet: "
+                , toText wid
+                ]
```
👍

Anviking committed:
Add goldens for pbkdf2 passwords

I manually tested these on current master — before the commits of this
PR — and they passed.

Anviking reviewed (approved):
LGTM. I suspect the `hashMaybe` we have on master wrong, but that's separate from this PR. I added a few goldens for PBKDF2 passwords as extra sanity check that this doesn't break anything.

rvl committed:
Use more hex in passphrase tests
